src/controller_examples/rabbitmq_controller/proof/helper_invariants/proof.rs

```diff
@@ -0,0 +1,1359 @@
+// Copyright 2022 VMware, Inc.
+// SPDX-License-Identifier: MIT
+#![allow(unused_imports)]
+use super::predicate::*;
+use crate::kubernetes_api_objects::{
+    api_method::*, common::*, error::*, owner_reference::*, prelude::*, resource::*,
+};
+use crate::kubernetes_cluster::spec::{
+    cluster::*,
+    cluster_state_machine::Step,
+    controller::common::{ControllerActionInput, ControllerStep},
+    message::*,
+};
+use crate::rabbitmq_controller::{
+    common::*,
+    proof::{predicate::*, resource::*},
+    spec::types::*,
+};
+use crate::temporal_logic::{defs::*, rules::*};
+use crate::vstd_ext::{multiset_lib, seq_lib, string_view::*};
+use vstd::{multiset::*, prelude::*, string::*};
+
+verus! {
+
+pub proof fn lemma_always_cr_objects_in_etcd_satisfy_state_validation(spec: TempPred<RMQCluster>)
+    requires
+        spec.entails(lift_state(RMQCluster::init())),
+        spec.entails(always(lift_action(RMQCluster::next()))),
+    ensures
+        spec.entails(always(lift_state(cr_objects_in_etcd_satisfy_state_validation()))),
+{
+    let inv = cr_objects_in_etcd_satisfy_state_validation();
+    RabbitmqClusterView::marshal_status_preserves_integrity();
+    init_invariant(spec, RMQCluster::init(), RMQCluster::next(), inv);
+}
+
+pub proof fn lemma_always_the_object_in_schedule_satisfies_state_validation(spec: TempPred<RMQCluster>)
+    requires
+        spec.entails(lift_state(RMQCluster::init())),
+        spec.entails(always(lift_action(RMQCluster::next()))),
+    ensures
+        spec.entails(always(lift_state(the_object_in_schedule_satisfies_state_validation()))),
+{
+    let inv = the_object_in_schedule_satisfies_state_validation();
+    let stronger_next = |s: RMQCluster, s_prime: RMQCluster| {
+        &&& RMQCluster::next()(s, s_prime)
+        &&& cr_objects_in_etcd_satisfy_state_validation()(s)
+    };
+    lemma_always_cr_objects_in_etcd_satisfy_state_validation(spec);
+    combine_spec_entails_always_n!(
+        spec, lift_action(stronger_next),
+        lift_action(RMQCluster::next()),
+        lift_state(cr_objects_in_etcd_satisfy_state_validation())
+    );
+    init_invariant(spec, RMQCluster::init(), stronger_next, inv);
+}
+
+pub proof fn lemma_always_the_object_in_reconcile_satisfies_state_validation(spec: TempPred<RMQCluster>)
```

Comment:
Same as above

Reply:
I just think we discussed this before...

Reply:
And I don't want to add those ugly pre conditions so I wrote them for rabbitmq controller

Reply:
Since these lemmas are pretty short, so it's also fine to copy them